tracing-subscriber/src/fmt/span.rs

```diff
@@ -236,26 +251,28 @@ impl Store {
     #[inline]
     pub(crate) fn current(&self) -> Option<Id> {
         CONTEXT
-            .try_with(|current| current.borrow().last().map(|span| self.clone_span(span)))
+            .try_with(|current| {
+                find_last_no_duplicate(&current.borrow()).map(|id| self.clone_span(id))
+            })
             .ok()?
     }
 
     pub(crate) fn push(&self, id: &Id) {
         let _ = CONTEXT.try_with(|current| {
             let mut current = current.borrow_mut();
-            if current.contains(id) {
-                // Ignore duplicate enters.
-                return;
-            }
-            current.push(self.clone_span(id));
+            let duplicate = current.iter().any(|context_id| id == &context_id.id);
```

Comment:
This, on the other hand, will probably do more iterating than the iteration in `find_last_no_duplicate` --- in most cases, we will have to iterate over the whole stack. 

As an alternative, we could consider a method of determining if an entered span ID is a duplicate without having to iterate. Here's one potential approach that might work:
1. We add a type like this (names, of course, TBD):
```rust
  struct SpanStack {
      stack: Vec<ContextId>,
      ids: HashSet<Id>,
  }
  ```
2. When pushing to a `SpanStack`, we can check if the ID is duplicated by seeing if it's in the `HashSet`, and insert it if it's a new ID:
 ```rust
  impl SpanStack {
      fn push(&mut self, id: Id) {
          let duplicate = self.ids.contains(&id);
          if !duplicate {
              self.ids.insert(id.clone());
          }
          self.stack.push(ContextId {
              duplicate,
              id
          });
      }
  
      fn pop(&mut self) -> Option<Id> {
          // we can also move the `pop` logic for only popping if 
          // we saw the expected span ID here, but I punted on that
          // because i just wanted to illustrate the proposed behaviour.
          let ContextId { id, duplicate } = self.stack.pop()?;
          if !duplicate {
              self.ids.remove(&id);
          }
          Some(id)
      }
  }
 ```
3. Replace the thread-local vec with a `SpanStack`.

This would give us an O(1) way of determining if a pushed ID is a duplicate, rather than having to iterate over the whole stack in the worst case. On the other hand, it is a bit more complex.

Reply:
good idea :)